### ISSUE TITLE
Can I use oneflow to efficiently implement methods like torch.nn.functional.unfold() and torch.nn.functional.fold()

### ISSUE THREAD

**Issue by 15168387900** [Oneflow-Inc/oneflow#3785]
I have implemented torch.nn.functional.unfold() and torch.nn.functional.fold() by myself using functions such as oneflow.pad(), oneflow.slice(), oneflow.reshape() and oneflow.concat() Similar methods, but in the testing process, they are really too inefficient.

**Comment by MARD1NO** [comment]
Thanks for your feedback. The `torch.nn.functional.fold` and unfold ops is still in developing process, it will be merged soon.

**Comment by 15168387900** [comment]
> Thanks for your feedback. The `torch.nn.functional.fold` and unfold ops is still in developing process, it will be merged soon.

Can you help me optimize my custom ops?
They run for several minutes, which is too slow.

`def unfold(name, input, kernel_size=3, padding=1, stride=1):
    x1 = flow.pad(input, paddings=((0, 0), (0, 0), (padding, padding), (padding, padding)))
    n, c, h, w = x1.shape
    l1 = math.ceil((h - kernel_size + 1) / stride)
    l2 = math.ceil((w - kernel_size + 1) / stride)
    output = flow.get_variable(
        name,
        dtype=flow.float32,
        shape=(n, c * kernel_size * kernel_size, 0),
        initializer=flow.zeros_initializer()
    )
    for i in range(0, l1):
        for j in range(0, l2):
            x2 = flow.slice(x1, begin=[None, None, i * stride, j * stride],
                            size=[None, None, kernel_size, kernel_size])
            x2 = flow.reshape(x2, shape=[n, c * kernel_size * kernel_size, 1])
            output = flow.concat(inputs=[output, x2], axis=2)
    return output`

`def fold(input, output_size, kernel_size, padding, stride):
    n = input.shape[0]
    c = input.shape[1] // kernel_size // kernel_size
    h = output_size[0] + 2 * padding
    w = output_size[1] + 2 * padding
    l1 = (h - kernel_size + 1) // stride
    l2 = (w - kernel_size + 1) // stride
    shape = [n, c, kernel_size, kernel_size]
    x1 = flow.slice(input, begin=[None, None, 0], size=[None, None, 1])
    x1 = flow.reshape(x1, shape)
    for i in range(1, l2):
        x11 = flow.slice(input, begin=[None, None, i], size=[None, None, 1])
        x11 = flow.reshape(x11, shape)
        x11 = flow.slice(x11, begin=[None, None, None, kernel_size - stride], size=[None, None, None, stride])
        x1 = flow.concat(inputs=[x1, x11], axis=3)
    for i in range(1, l1):
        x2 = flow.slice(input, begin=[None, None, i * l2], size=[None, None, 1])
        x2 = flow.reshape(x2, shape)
        x2 = flow.slice(x2, begin=[None, None, kernel_size - stride, None], size=[None, None, stride, None])
        for j in range(1, l2):
            x22 = flow.slice(input, begin=[None, None, i * l2 + j], size=[None, None, 1])
            x22 = flow.reshape(x22, shape)
            x22 = flow.slice(x22, begin=[None, None, kernel_size - stride, kernel_size - stride], size=[None, None, stride, stride])
            x2 = flow.concat(inputs=[x2, x22], axis=3)
        x1 = flow.concat(inputs=[x1, x2], axis=2)
    output = flow.slice(x1, begin=[None, None, padding, padding], size=[None, None, output_size[0], output_size[1]])
    return output`

**Comment by 15168387900** [comment]
> Thanks for your feedback. The `torch.nn.functional.fold` and unfold ops is still in developing process, it will be merged soon.

Can you help me optimize my custom ops?
They run for several minutes, which is too slow : (

```
def unfold(name, input, kernel_size=3, padding=1, stride=1):
    x1 = flow.pad(input, paddings=((0, 0), (0, 0), (padding, padding), (padding, padding)))
    n, c, h, w = x1.shape
    l1 = math.ceil((h - kernel_size + 1) / stride)
    l2 = math.ceil((w - kernel_size + 1) / stride)
    output = flow.get_variable(
        name,
        dtype=flow.float32,
        shape=(n, c * kernel_size * kernel_size, 0),
        initializer=flow.zeros_initializer()
    )
    for i in range(0, l1):
        for j in range(0, l2):
            x2 = flow.slice(x1, begin=[None, None, i * stride, j * stride],
                            size=[None, None, kernel_size, kernel_size])
            x2 = flow.reshape(x2, shape=[n, c * kernel_size * kernel_size, 1])
            output = flow.concat(inputs=[output, x2], axis=2)
    return output
```

```
def fold(input, output_size, kernel_size, padding, stride):
    n = input.shape[0]
    c = input.shape[1] // kernel_size // kernel_size
    h = output_size[0] + 2 * padding
    w = output_size[1] + 2 * padding
    l1 = (h - kernel_size + 1) // stride
    l2 = (w - kernel_size + 1) // stride
    shape = [n, c, kernel_size, kernel_size]
    x1 = flow.slice(input, begin=[None, None, 0], size=[None, None, 1])
    x1 = flow.reshape(x1, shape)
    for i in range(1, l2):
        x11 = flow.slice(input, begin=[None, None, i], size=[None, None, 1])
        x11 = flow.reshape(x11, shape)
        x11 = flow.slice(x11, begin=[None, None, None, kernel_size - stride], size=[None, None, None, stride])
        x1 = flow.concat(inputs=[x1, x11], axis=3)
    for i in range(1, l1):
        x2 = flow.slice(input, begin=[None, None, i * l2], size=[None, None, 1])
        x2 = flow.reshape(x2, shape)
        x2 = flow.slice(x2, begin=[None, None, kernel_size - stride, None], size=[None, None, stride, None])
        for j in range(1, l2):
            x22 = flow.slice(input, begin=[None, None, i * l2 + j], size=[None, None, 1])
            x22 = flow.reshape(x22, shape)
            x22 = flow.slice(x22, begin=[None, None, kernel_size - stride, kernel_size - stride], size=[None, None, stride, stride])
            x2 = flow.concat(inputs=[x2, x22], axis=3)
        x1 = flow.concat(inputs=[x1, x2], axis=2)
    output = flow.slice(x1, begin=[None, None, padding, padding], size=[None, None, output_size[0], output_size[1]])
    return output
```

**Comment by doombeaker** [comment]
`unfold` is under developing(https://github.com/Oneflow-Inc/oneflow/pull/3827), but sorry for that it may NOT be merged soon for the reason of code quality.

`fold` ops will open a PR after `unfold` merged.

You can use python warpper instead now to make program run, for efficiently, C++ implemention is needed. Or if you are intrested in implentation this op and contribute it to oneflow, we are glad to offer most assistatns.

**Comment by kaijieshi7** [comment]
到现在还没有解决吗？
